### PR TITLE
WIP: check that records use the correct syntax at parsing time.

### PR DESCRIPTION
**Pull request by Zimmi48**
This is a super-preliminary refactoring of the parsing of `Inductive` / `Record` following a suggestion of @jfehrle. In practice, the parser is not able to reject the combination of `Record` / `Structure` + the constructor syntax, does not know that the `@binders | @binders` syntax for uniform parameters  is not supported either, does not know that `Variant` does not support the record syntax. `Class` piggies-back on the constructor syntax for definitional classes, but rejects definitions with multiple constructors and it rejects mutual definitions. Finally, `Record` / `Structure` do support the `with` construct but I'm not sure with what use case since any properly (co-)recursive record must be constructed with (Co)Inductive.

Note that once there is more separation at parsing time, this could be reflected in the datatypes in `vernacexpr.ml` and eventually trickle down, leading to removal of some error paths. For instance, currently, at parsing time we could distinguish between definitional and non-definitional classes, but we don't do it. We give a dummy value to the constructor `Class` and we fix this value later, in `vernacentries.ml`.

Note that even this PR turns out to lead nowhere or does not raise consensus, I still think that we must do this kind of refactoring in the grammar that we document.

### CODE REVIEW

**Comment by jfehrle** [comment]
If non-Record/Structure definitions can't use `record_fields` then the two `record_field` productions here should be removed, too.  And then the nonterminal name can be simplified.

```
constructor_list_or_record_decl: [
| "|" LIST1 constructor SEP "|"
| identref constructor_type "|" LIST1 constructor SEP "|"
| identref constructor_type
| identref "{" record_fields "}"
| "{" record_fields "}"
|
]
```

File: vernac/g_vernac.mlg [hunk 1]
```diff
@@ -409,6 +413,13 @@ GRAMMAR EXTEND Gram
       | "{";fs = record_fields; "}" -> { RecordDecl (None,fs) }
       |  -> { Constructors [] } ] ]
   ;
+  record_definition:
+    [ [ oc = opt_coercion; id = ident_decl; indpar = binders;
```
Comment (jfehrle) [inline vernac/g_vernac.mlg]: Just to be clear, note that this drops the `OPT [ "|" binders ]` included in `inductive_definition`.  This clause doesn't apply to records?

Comment (Zimmi48) [reply vernac/g_vernac.mlg]: Indeed, it doesn't: 

```
Record foo (n : nat) | (m : nat) := {}.
(* Error: Records do not support the "|" syntax. *)
```

**Comment by Zimmi48** [comment]
A funny example of where this leads is the following:

```
Variant foo.
(* Error: The Variant keyword does not support syntax { ... }. *)
```

**Comment by Zimmi48** [comment]
I won't rebase this further. Deferring to @herbelin to decide if this is something worth pursuing or not.

**Comment by herbelin** [comment]
Taking over.

**Comment by herbelin** [comment]
I went up to adding a new `VernacRecord` constructor. I'm however now going to make a break on it concentrating first on making the code base interpreting records and the code base interpreting inductive types closer (see discussion at [15489](https://github.com/coq/coq/pull/15489#issuecomment-1040914531)).

Note that doc_gram is not updated yet. It will be at some time, when things stabilize.

**Pull request closed without merging**